sound/soc/intel/boards/tgl_rt1308.c

```diff
@@ -0,0 +1,331 @@
+		.dai_fmt	= SND_SOC_DAIFMT_I2S | SND_SOC_DAIFMT_NB_NF |
+				  SND_SOC_DAIFMT_CBS_CFS,
+		.dpcm_playback = 1,
+		.nonatomic = true,
+		SND_SOC_DAILINK_REG(ssp1_pin, ssp1_codec, platform),
+	},
+	{
+		.name = "dmic01",
+		.id = 1,
+		.ignore_suspend = 1,
+		.dpcm_capture = 1,
+		.no_pcm = 1,
+		SND_SOC_DAILINK_REG(dmic_pin, dmic_codec, platform),
+	},
+#if IS_ENABLED(CONFIG_SND_SOC_HDAC_HDMI)
+	{
+		.name = "iDisp1",
+		.id = 2,
+		.init = tgl_hdmi_init,
+		.dpcm_playback = 1,
+		.no_pcm = 1,
+		SND_SOC_DAILINK_REG(idisp1_pin, idisp1_codec, platform),
+	},
+	{
+		.name = "iDisp2",
+		.id = 3,
+		.init = tgl_hdmi_init,
+		.dpcm_playback = 1,
+		.no_pcm = 1,
+		SND_SOC_DAILINK_REG(idisp2_pin, idisp2_codec, platform),
+	},
+	{
+		.name = "iDisp3",
+		.id = 4,
+		.init = tgl_hdmi_init,
+		.dpcm_playback = 1,
+		.no_pcm = 1,
+		SND_SOC_DAILINK_REG(idisp3_pin, idisp3_codec, platform),
+	},
+	{
+		.name = "iDisp4",
+		.id = 5,
+		.init = tgl_hdmi_init,
+		.dpcm_playback = 1,
+		.no_pcm = 1,
+		SND_SOC_DAILINK_REG(idisp4_pin, idisp4_codec, platform),
+	},
+#endif
+};
+
+/* audio machine driver */
+static struct snd_soc_card tgl_rt1308_card = {
+	.name         = "tgl_rt1308",
+	.owner        = THIS_MODULE,
+	.dai_link     = tgl_rt1308_dailink,
+	.num_links = ARRAY_SIZE(tgl_rt1308_dailink),
+	.controls = tgl_rt1308_controls,
+	.num_controls = ARRAY_SIZE(tgl_rt1308_controls),
+	.dapm_widgets = tgl_rt1308_dapm_widgets,
+	.num_dapm_widgets = ARRAY_SIZE(tgl_rt1308_dapm_widgets),
+	.dapm_routes = tgl_rt1308_dapm_routes,
+	.num_dapm_routes = ARRAY_SIZE(tgl_rt1308_dapm_routes),
+	.late_probe = tgl_card_late_probe,
+};
+
+static int tgl_rt1308_probe(struct platform_device *pdev)
+{
+	struct snd_soc_acpi_mach *mach;
+	struct tgl_card_private *ctx;
+	struct snd_soc_card *card;
+	int ret = 0;
+
+	card = &tgl_rt1308_card;
+	card->dev = &pdev->dev;
+
+	ctx = devm_kzalloc(&pdev->dev, sizeof(*ctx), GFP_KERNEL);
+	if (!ctx)
+		return -ENOMEM;
+
+	if (IS_ENABLED(CONFIG_SND_SOC_HDAC_HDMI))
+		INIT_LIST_HEAD(&ctx->hdmi_pcm_list);
+
+	tgl_rt1308_card.dev = &pdev->dev;
+	mach = (&pdev->dev)->platform_data;
+
+	ret = snd_soc_fixup_dai_links_platform_name(&tgl_rt1308_card,
+						    mach->mach_params.platform);
+	if (ret)
+		return ret;
+
+	snd_soc_card_set_drvdata(card, ctx);
+
+	ret = devm_snd_soc_register_card(&pdev->dev,
+					  &tgl_rt1308_card);
+	if (ret) {
+		dev_err(&pdev->dev,
+			"snd_soc_register_card failed %d\n", ret);
+		return ret;
+	}
+
+	platform_set_drvdata(pdev, card);
+
+	return ret;
+
+}
+
+static int tgl_rt1308_remove(struct platform_device *pdev)
+{
+	return snd_soc_unregister_card(&tgl_rt1308_card);
+}
+
+static struct platform_driver tgl_rt1308_driver = {
+	.driver = {
+		.name   = "tgl_rt1308",
+		.owner  = THIS_MODULE,
+	},
+	.probe          = tgl_rt1308_probe,
+	.remove         = tgl_rt1308_remove,
+};
+
+module_platform_driver(tgl_rt1308_driver);
+
+MODULE_AUTHOR("Xiuli Pan");
+MODULE_DESCRIPTION("ASoC Intel(R) Tigerlak + RT1308 Machine driver");
```

Comment:
typo: s/Tigerlak/Tigerlake/
And the more official way to write the name would be actually "Tiger Lake", but it seems we've used the Foobarlake convention elsewhere already, so ok with this.